src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,449 @@
+
+lemma ideal.homogeneous_core_eq_Sup :
+  I.homogeneous_core' 𝒜 = Sup {J : ideal A | J.is_homogeneous 𝒜 ∧ J ≤ I} :=
+begin
+  refine (is_lub.Sup_eq _).symm,
+  apply is_greatest.is_lub,
+  have coe_mono : monotone (coe : {I : ideal A // I.is_homogeneous 𝒜} → ideal A) := λ _ _, id,
+  convert coe_mono.map_is_greatest (ideal.homogeneous_core.gc 𝒜).is_greatest_u using 1,
+  simp only [subtype.coe_image, exists_prop, mem_set_of_eq, subtype.coe_mk],
+end
+
+end homogeneous_core
+
+section homogeneous_hull
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A]
+variables [algebra R A] [decidable_eq ι] [add_comm_monoid ι]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+variable (I : ideal A)
+
+/--For any `I : ideal R`, not necessarily homogeneous, `I.homogeneous_hull' 𝒜` is
+the smallest homogeneous ideal containing `I`.-/
+def ideal.homogeneous_hull : homogeneous_ideal 𝒜 :=
+⟨ideal.span {r : A | ∃ (i : ι) (x : I), (graded_algebra.decompose 𝒜 x i : A) = r}, begin
+  rw ideal.is_homogeneous.iff_exists,
+  use {x : homogeneous_submonoid 𝒜 | ∃ (i : ι) (r : I), (graded_algebra.decompose 𝒜 r i : A) = x},
+  congr,
+  ext r,
+  split;
+  intros h,
+  { obtain ⟨i, ⟨x, hx1⟩, hx2⟩ := h,
+    exact ⟨⟨_, is_homogeneous_coe _⟩, ⟨⟨i, ⟨⟨x, hx1⟩, rfl⟩⟩, hx2⟩⟩,},
+  { obtain ⟨_, ⟨⟨i, ⟨⟨r, hr⟩, h⟩⟩, rfl⟩⟩ := h,
+    use i, use ⟨r, hr⟩, exact h }
+end⟩
+
+lemma ideal.ideal_le_homogeneous_hull :
+  I ≤ ideal.homogeneous_hull 𝒜 I :=
+begin
+  intros r hr,
+  letI : Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  rw [←graded_algebra.sum_support_decompose 𝒜 r],
+  refine ideal.sum_mem _ _, intros j hj,
+  apply ideal.subset_span, use j, use ⟨r, hr⟩, refl,
+end
+
+lemma ideal.homogeneous_hull_is_mono : monotone (ideal.homogeneous_hull 𝒜) := λ I J I_le_J,
+begin
+  apply ideal.span_mono,
+  rintros r ⟨hr1, ⟨x, hx⟩, rfl⟩,
+  refine ⟨hr1, ⟨⟨x, I_le_J hx⟩, rfl⟩⟩,
+end
+
+lemma ideal.homogeneous_hull_eq_Inf :
+  ideal.homogeneous_hull 𝒜 I = Inf { J : homogeneous_ideal 𝒜 | I ≤ J } :=
+begin
+  ext,
+  split;
+  intros hx,
+  { erw ideal.mem_Inf,
+    rintros _ ⟨K, HK1, rfl⟩,
+    erw [ideal.mem_span] at hx,
+    apply hx K,
+    rintros r ⟨i, ⟨⟨y, hy⟩, rfl⟩⟩,
+    exact K.2 _ (HK1 hy), },
+  { erw ideal.mem_Inf at hx,
+    refine @hx (ideal.homogeneous_hull 𝒜 I) _,
+    exact ⟨ideal.homogeneous_hull _ _, ideal.ideal_le_homogeneous_hull _ _, rfl⟩, }
+end
+
+lemma homogeneous_hull_eq_supr :
+   ↑(I.homogeneous_hull 𝒜) = ⨆ i, ideal.span (graded_algebra.proj 𝒜 i '' I) :=
+ begin
+   rw ←ideal.span_Union,
+   apply congr_arg ideal.span _,
+   ext1,
+   simp only [set.mem_Union, set.mem_image, mem_set_of_eq, graded_algebra.proj_apply,
+     set_like.exists, exists_prop, subtype.coe_mk, set_like.mem_coe],
+ end
+
+lemma homogeneous_hull_eq_supr' :
+  (I.homogeneous_hull 𝒜) =
+  ⨆ i, ⟨ideal.span (graded_algebra.proj 𝒜 i '' I), ideal.is_homogeneous_span 𝒜 _ (begin
+    rintros _ ⟨x, h, rfl⟩,
+    refine ⟨i, submodule.coe_mem _⟩,
```

Comment:
Does `set_like.is_homogeneous_coe` make this proof shorter somehow?

Reply:
yes
```lean
lemma homogeneous_hull_eq_supr :
  (I.homogeneous_hull 𝒜) =
  ⨆ i, ⟨ideal.span (graded_algebra.proj 𝒜 i '' I), ideal.is_homogeneous_span 𝒜 _ 
    (by {rintros _ ⟨x, h, rfl⟩, apply set_like.is_homogeneous_coe})⟩ :=
by { ext1, rw [coe_homogeneous_hull_eq_supr, homogeneous_ideal.coe_supr], refl, }
```